File: src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/DNDecommissionStatusCheckerTask.java
package com.sequenceiq.cloudbreak.service.cluster.flow;

import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import com.sequenceiq.ambari.client.AmbariClient;
import com.sequenceiq.cloudbreak.service.StackBasedStatusCheckerTask;

@Component
public class DNDecommissionStatusCheckerTask extends StackBasedStatusCheckerTask<AmbariOperations> {

    private static final Logger LOGGER = LoggerFactory.getLogger(DNDecommissionStatusCheckerTask.class);

    @Override
    public boolean checkStatus(AmbariOperations t) {
        AmbariClient ambariClient = t.getAmbariClient();
        Map<String, Long> dataNodes = ambariClient.getDecommissioningDataNodes();
        boolean finished = dataNodes.isEmpty();
        if (!finished) {
            LOGGER.info("DataNode decommission is in progress: {}", dataNodes);
        }
        return finished;
    }

    @Override
    public void handleTimeout(AmbariOperations t) {
        throw new IllegalStateException("DataNode decommission timed out");

    }

    @Override
    public String successMessage(AmbariOperations t) {
        return "Requested DataNode decommission operations completed";
    }

}

File: core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/DNDecommissionStatusCheckerTask.kt
package com.sequenceiq.cloudbreak.service.cluster.flow

import com.sequenceiq.ambari.client.AmbariClient
import com.sequenceiq.cloudbreak.service.ClusterBasedStatusCheckerTask
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class DNDecommissionStatusCheckerTask : ClusterBasedStatusCheckerTask<AmbariOperations>() {

    override fun checkStatus(t: AmbariOperations): Boolean {
        val ambariClient = t.ambariClient
        val dataNodes = ambariClient.decommissioningDataNodes
        val finished = dataNodes.isEmpty()
        if (!finished) {
            LOGGER.info("DataNode decommission is in progress: {}", dataNodes)
        }
        return finished
    }

    override fun handleTimeout(t: AmbariOperations) {
        throw IllegalStateException("DataNode decommission timed out")

    }

    override fun successMessage(t: AmbariOperations): String {
        return "Requested DataNode decommission operations completed"
    }

    companion object {

        private val LOGGER = LoggerFactory.getLogger(DNDecommissionStatusCheckerTask::class.java)
    }

}
